### PR TITLE
[Core] RGB matrix ws2812 update

### PR DESCRIPTION
**Pull request by Xelus22**
## Description

Goes with https://github.com/qmk/qmk_firmware/pull/21134 . This change makes it so that the WS2812 doesn't update every cycle if no RGB LEDs have updated colour. (mainly for static and bitbang applications)

**Untested because I don't have any WS2812 only rgb matrix boards**

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* WS2812 don't update every RGB Matrix cycle

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/rgb_matrix/rgb_matrix_drivers.c [hunk 1]
```diff
@@ -426,11 +426,17 @@ const rgb_matrix_driver_t rgb_matrix_driver = {
 
 // LED color buffer
 LED_TYPE rgb_matrix_ws2812_array[RGB_MATRIX_LED_COUNT];
+bool updateLEDs = false;
```
Comment (drashna) [inline quantum/rgb_matrix/rgb_matrix_drivers.c:429]:
```suggestion
bool     updateLEDs = false;
```

File: quantum/rgb_matrix/rgb_matrix_drivers.c [hunk 2]
```diff
@@ -448,6 +454,13 @@ static inline void setled(int i, uint8_t r, uint8_t g, uint8_t b) {
     }
 #    endif
 
+    if (rgb_matrix_ws2812_array[i].r == r &&
+        rgb_matrix_ws2812_array[i].g == g &&
+        rgb_matrix_ws2812_array[i].b == b){
+        return;
+    }
+
+    updateLEDs = true;
```
Comment (drashna) [inline quantum/rgb_matrix/rgb_matrix_drivers.c:463]: And: 
```suggestion
    if (rgb_matrix_ws2812_array[i].r == r && rgb_matrix_ws2812_array[i].g == g && rgb_matrix_ws2812_array[i].b == b) {
        return;
    }

    updateLEDs                   = true;
```

**Review by drashna — COMMENTED** [review]
Lint angry.

Otherwise tested, and works well.

**Comment by ToastyStoemp** [comment]
Also tested this on Atmega32u4 PCB (126 WS2812 leds)
Matrix scan frequency:

- Master branch: 

1. Off							789
2. Solid						650
3. Gradient Up/Down	424
4. Cycle Left/Right		418
5. Cycle Spiral				286

- Xelus WS2812 update branch:

1. Off							799
2. Solid						733
3. Gradient Up/Down	453
4. Cycle Left/Right		414
5. Cycle Spiral				284

**Review by drashna — APPROVED** [review]

File: quantum/rgb_matrix/rgb_matrix_drivers.c [hunk 3]
```diff
@@ -426,11 +426,17 @@ const rgb_matrix_driver_t rgb_matrix_driver = {
 
 // LED color buffer
 LED_TYPE rgb_matrix_ws2812_array[RGB_MATRIX_LED_COUNT];
+bool     updateLEDs = false;
```
Comment (zvecr) [inline quantum/rgb_matrix/rgb_matrix_drivers.c:429]: Jumping from a snake case variable to `updateLEDs` is a little inconsistent.

Should probably be called something like `ws2812_dirty` or `update_required`.

File: quantum/rgb_matrix/rgb_matrix_drivers.c [hunk 4]
```diff
@@ -426,11 +426,17 @@ const rgb_matrix_driver_t rgb_matrix_driver = {
 
 // LED color buffer
 LED_TYPE rgb_matrix_ws2812_array[RGB_MATRIX_LED_COUNT];
+bool ws2812_dirty = false;
```
Comment (zvecr) [inline quantum/rgb_matrix/rgb_matrix_drivers.c:429]:
```suggestion
bool     ws2812_dirty = false;
```

**Review by zvecr — CHANGES_REQUESTED** [review]
Lint fixes

File: quantum/rgb_matrix/rgb_matrix_drivers.c [hunk 5]
```diff
@@ -448,6 +454,11 @@ static inline void setled(int i, uint8_t r, uint8_t g, uint8_t b) {
     }
 #    endif
 
+    if (rgb_matrix_ws2812_array[i].r == r && rgb_matrix_ws2812_array[i].g == g && rgb_matrix_ws2812_array[i].b == b) {
+        return;
+    }
+
+    ws2812_dirty = true;
```
Comment (zvecr) [inline quantum/rgb_matrix/rgb_matrix_drivers.c:461]:
```suggestion
    ws2812_dirty                 = true;
```

**Review by zvecr — APPROVED** [review]

**Pull request merged**